### ISSUE TITLE
Focal Loss for Dense Object Detection

### ISSUE THREAD

**Issue by Yagami360** [Yagami360/MachineLearning-Papers_Survey#66]
## 0. 論文情報・リンク
- 論文リンク：https://arxiv.org/abs/1708.02002
- 公開日時：2017/08/07
- 被引用数（記事作成時点）：3426件
- 実装コード（公式）: https://github.com/facebookresearch/Detectron
    ※ Mask R-CNN と RetinaNet のコード。focal loss のコードは、loss 値定義あたりに含まれていると思われる
- 実装コード（PyTorch）：https://github.com/clcarwin/focal_loss_pytorch
- Publication : 

## 1. どんなもの？
- １ステージの物体検出器において、背景クラスと前景クラスとの間のクラス間不均衡（背景のほうが圧倒的に領域がでかい）がある場合でもうまく学習出来るように cross entopy loss を再構成した focal loss を提案し、１ステージ物体検出器でも２ステージ物体検出器の精度を上回ることを実現
    <image src="https://user-images.githubusercontent.com/25688193/85203008-9c065f00-b345-11ea-9881-bf7b021ccc7b.png" width="300">

## 2. 先行研究と比べてどこがすごいの？
- R-CNN やそれらを世襲した物体検出モデルは、２ステージで構成される物体検出器であり、高精度の物体検出が出来る代わりに推論速度は低速であるという問題が存在する。
	一方 YOLO や SSD などの１ステージで構成される物体検出器は、推論速度は高速であるが、物体検出の精度は低くなってしまう。
	本手法では、１ステージ物体検出器での検出精度が低下してしまうのは背景クラスと前景クラスとの間のクラス間不均衡が原因であるという観点から、それを解決するために cross entopy loss を再構成した focal loss を提案し、１ステージ物体検出器でも２ステージ検出器の精度を上回ることを実現している。

## 3. 技術や手法の"キモ"はどこにある？
- Focal Loss
	R-CNN などの２ステージ物体検出器では、two-stage cascade とsampling heuristics により、背景クラスと前景クラスとの間のクラス間不均衡（背景のほうが圧倒的に領域がでかい）に対処している。一方 YOLO や SSD などの１ステージ物体検出器は、推論速度は高速である一方で、このようなクラス間不均衡への対処が行われていないために、検出精度が低下してしまうという問題が存在する。
	
	１ステージ物体検出器においてクラス間不均衡に対処する方法としては、 bootstrapping や hard example mining が考えられるが、これらの手法は処理効率が悪いという問題がある。
	
	本手法ではこの問題を解決するために、１ステージ物体検出器においてより効率的にクラス間不均衡に対処できる Focal Loss を提案している。

    <image src="https://user-images.githubusercontent.com/25688193/85203025-c1936880-b345-11ea-947a-2b8b3e2a600d.png" width="320">

	上図は、cross entropy loss（青線）とFocal Loss（青線以外）を示した図である。
	青線を見てわかるように、cross entropy loss ではモデルの推論確率 p が半々の確率よりも圧倒的に高く確信度の高い状態（p≫0.5）でも、損失関数は非ゼロになっている。
	従って、多数のサンプル点で cross entropy loss の和を取るときに、この小さな loss 値の合計は、確信度が低く（p≪0.5）損失関数値が大きいクラスの値を凌駕してしまうケースが発生する。
	そして、その結果としてクラス間不均衡がある場合に物体検出精度が低下してしまう。
	
	Focal Loss では、この問題に対処するために、正しいクラスの確信度が高まるにつれて（p≫0.5）、スケーリング係数をゼロに減衰させ、損失関数値がより小さくなるようにクロスエントロピー損失関数を再構成する。
	式で書くと、以下のような式となる。
    <image src="https://user-images.githubusercontent.com/25688193/85203042-d839bf80-b345-11ea-8439-461724685974.png" width="420">

## 4. どうやって有効だと検証した？
- 既存の物体検出モデルとの定量比較
    <image src="https://user-images.githubusercontent.com/25688193/85203071-fd2e3280-b345-11ea-87e5-2348d79e7d1a.png" width="325">
    <image src="https://user-images.githubusercontent.com/25688193/85203081-0b7c4e80-b346-11ea-8fa7-10d2a7dbc9f7.png" width="620">
	focal loss を有効化した RetineNet では、単純化した１ステージ検出器であるにも関わらず、AP スコアが既存のモデル（１ステージ検出器及び２ステージ検出器の両方）を上回ってており、高精度の物体検出を行えていることが見て取れる。

   - 【補足】RetinaNet のアーキテクチャ
        <image src="https://user-images.githubusercontent.com/25688193/85203090-1cc55b00-b346-11ea-9fb8-4b4f7b596237.png" width="570">
        focal loss の有用性を確認するために設計された物体検出モデルで、単純化された１ステージ物体検出器になっている

## 5. 議論はあるか？
- xxx

## 6. 次に読むべき論文はあるか？
- xxx

## 7. 参考文献
- xxx
